### DISCUSSION TITLE
Help me with lua code

### DISCUSSION THREAD

**Discussion by SmallBunny6** [community/community#134126] [Programming Help]
### Body

-- Services
local TweenService = game:GetService("TweenService")
local Players = game:GetService("Players")
local MarketplaceService = game:GetService("MarketplaceService")

-- References
local tycoon = script.Parent.Parent
local MainItems = tycoon:FindFirstChild("MainItems")
local values = tycoon:FindFirstChild("Values")
local buttons = tycoon:FindFirstChild("Buttons")
local PurchasedItems = tycoon:FindFirstChild("PurchasedItems")
local audio = tycoon:FindFirstChild("Audio")

-- Debugowanie odniesień
if not MainItems then
	error("MainItems nie istnieje!")
end

if not MainItems:FindFirstChild("OwnerDoor") then
	error("OwnerDoor nie istnieje w MainItems!")
end

-- Variables
local objects = {}
local random = Random.new()
local BUILDING_ANIMATION_POSITION_OFFSET_AMOUNT = 2
local BUILDING_ANIMATION_PART_DELAY = 0.03

-- Helper functions
function hasProperty(instance, property)
	assert(typeof(instance) == "Instance")
	assert(typeof(property) == "string")

	local hasProperty = false
	pcall(function()
		local v = instance[property]
		hasProperty = true 
	end)

	return hasProperty
end

function instanceListToPropertyDict(instances, propertyList)
	assert(typeof(instances) == "table")
	assert(typeof(propertyList) == "table")

	local dict = {}
	for _, instance in ipairs(instances) do
		local dictEntry = {}
		for _, property in pairs(propertyList) do
			assert(hasProperty(instance, property), string.format("Instance '%s' (a %s) doesn't have property '%s'.", tostring(instance), instance.ClassName, property))
			dictEntry[property] = instance[property]
		end
		dict[instance] = dictEntry
	end
	return dict
end

function getDescendantsWhichAre(ancestor, className)
	assert(typeof(ancestor) == "Instance")
	assert(typeof(className) == "string")

	local descendants = {}
	for _, descendant in pairs(ancestor:GetDescendants()) do
		if descendant:IsA(className) then
			table.insert(descendants, descendant)
		end
	end
	return descendants
end

function animateBuildingIn(buildingModel, tweenInfo)
	assert(typeof(buildingModel) == "Instance" and buildingModel.ClassName == "Model", string.format("Invalid argument #1 to 'animateBuildingIn' (Model expected, got %s)", typeof(buildingModel)))
	assert(typeof(tweenInfo) == "TweenInfo", string.format("Invalid argument #2 to 'animateBuildingIn' (TweenInfo expected, got %s)", typeof(tweenInfo)))

	local parts = getDescendantsWhichAre(buildingModel, "BasePart")
	local originalProperties = instanceListToPropertyDict(parts, {"Transparency", "CFrame", "Color"})

	for _, part in pairs(parts) do
		part.Transparency = 1
		part.Color = Color3.fromRGB(255, 255, 255)

		local positionOffset = Vector3.new(random:NextNumber(-1, 1), random:NextNumber(-0.25, 1.75), random:NextNumber(-1, 1)) * BUILDING_ANIMATION_POSITION_OFFSET_AMOUNT
		part.CFrame = part.CFrame * CFrame.new(positionOffset)
	end

	for _, part in pairs(parts) do
		local tween = TweenService:Create(part, tweenInfo, originalProperties[part])

		tween.Completed:Connect(function()
			part.Transparency = originalProperties[part].Transparency
			part.CFrame = originalProperties[part].CFrame
		end)

		tween:Play()
		wait(BUILDING_ANIMATION_PART_DELAY)
	end
end

-- Main Logic - Tycoon Owner Door Functionality
MainItems.OwnerDoor.Door.Touched:Connect(function(hit)
	local player = Players:GetPlayerFromCharacter(hit.Parent)

	if player and values.OwnerValue.Value == nil then 
		if not player:FindFirstChild("HasTycoon") or player.HasTycoon.Value == false then 
			values.OwnerValue.Value = player 
			MainItems.OwnerDoor.Title.SurfaceGui.TextLabel.Text = tostring(values.OwnerValue.Value.Name) .. "'s Tycoon"
			player:SetAttribute("HasTycoon", true)
			print("Gracz stał się właścicielem Tycoona:", player.Name)  -- Debugging line 
		else 
			print("Gracz już posiada Tycoona.")
		end 
	else 
		print("Hit nie jest graczem lub Tycoon już ma właściciela.")
	end 
end)

if buttons then 
	for _, v in pairs(buttons:GetChildren()) do 
		if v:FindFirstChild("Button") then 
			local objectName = v.Object.Value 
			local priceValue = v.Price.Value

			spawn(function() 
				print("Przetwarzanie przycisku:", v.Name)  -- Debugging line 

				local newObject = PurchasedItems:FindFirstChild(objectName)
				if newObject then 
					objects[objectName] = newObject:Clone()
					newObject:Destroy() 
					print("Skopiowany obiekt:", objectName)  -- Debugging line  
				else  
					print("Nie znaleziono obiektu dla:", objectName)  -- Debugging line 
					v:Destroy() 
				end 

				if v:FindFirstChild("Dependency") then 
					v.Button.Transparency = 1 
					v.Button.CanCollide = false 

					coroutine.resume(coroutine.create(function() 
						if PurchasedItems:FindFirstChild(v.Dependency.Value) then 
							v.Button.Transparency = 0 
							v.Button.CanCollide = true 
						end 
					end))  
				end  

				v.Button.Touched:Connect(function(hit) 
					print("Przycisk dotknięty")  -- Debugging line 

					local player = Players:GetPlayerFromCharacter(hit.Parent)

					if player and values.OwnerValue.Value == player then 

						if objects[objectName] then  
							print("Gracz jest właścicielem")  -- Debugging line 

							if v.Button.CanCollide then  
								if player:FindFirstChild("LeaderStats") and player.LeaderStats.Cash.Value >= priceValue then  
									player.LeaderStats.Cash.Value = player.LeaderStats.Cash.Value - priceValue  

									-- Klonowanie obiektu Droppera i ustawianie go jako widocznego
									local clonedObject = objects[objectName]:Clone()  
									clonedObject.Parent = PurchasedItems  

									-- Ustawienie pozycji Droppera na położenie przycisku (zmień wartości według potrzeb)
									clonedObject:SetPrimaryPartCFrame(v.Button.Position + Vector3.new(0, v.Button.Size.Y / 2 + clonedObject.PrimaryPart.Size.Y / 2 + 1, 0)) -- Umieść Droppera nad przyciskiem

									-- Ustawienia widoczności Droppera
									clonedObject.PrimaryPart.Transparency = 0 -- Ustaw przezroczystość na 0 (widoczny)
									clonedObject.PrimaryPart.CanCollide = true -- Upewnij się, że koliduje

									print("Zakupiono:", objectName)  -- Debugging line 

									animateBuildingIn(clonedObject, TweenInfo.new(1, Enum.EasingStyle.Quad, Enum.EasingDirection.InOut))

									-- Pokaż przycisk kolorowania po zakupie Droppera i ukryj przycisk zakupu Droppera.
									local colorizerButton = MainItems:FindFirstChild("ColorizerButton")
									if colorizerButton then
										colorizerButton.Visible = true  
									end

									-- Ukryj przycisk zakupu po pierwszym zakupie.
									v.Button.Visible = false  
								else  
									print("Niewystarczająca ilość pieniędzy lub brak LeaderStats")  -- Debugging line  
								end  
							else  
								print("Przycisk nie może kolidować")  -- Debugging line  
							end  
						else  
							print("Dropper już zakupiony.") -- Informacja o przekroczeniu limitu zakupu Droppera.
						end  
					else  
						print("Gracz nie jest właścicielem lub hit jest nil.")  -- Debugging line  
					end   
				end)   
			end)   
		end   
	end   
end   

PurchasedItems.ChildAdded:Connect(function(child)
	if child:IsA("Model") then  
		print("Dziecko dodane do PurchasedItems:", child.Name)  -- Debugging line  

		task.delay(0.5, function() 
			animateBuildingIn(child, TweenInfo.new(1, Enum.EasingStyle.Quad, Enum.EasingDirection.InOut)) 
		end)
	end
end)
  Here is my code and OwnerDoor doesnt work and there another issue
    Workspace.Tycoons.Tycoon.Buttons.ColorizerButton.Button.Script:16: MainItems nie istnieje! 
    
![image](https://github.com/user-attachments/assets/dee868ee-3170-46f4-86f6-00a6e641b25b)
pls help me 

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by aradfarahani** [comment]
It looks like there are a couple of issues with your script:

### 1. **OwnerDoor Not Working**
The `OwnerDoor` functionality might not be working due to the following reasons:
- **Incorrect Path**: Ensure that the path to `OwnerDoor` is correct. Double-check the hierarchy in your workspace.
- **Event Connection**: Verify that the `Touched` event is correctly connected and that the `hit` parameter is being correctly processed.

### 2. **Error: `MainItems nie istnieje!`**
This error indicates that the `MainItems` folder is not found. Here are some steps to troubleshoot:

- **Check Hierarchy**: Ensure that `MainItems` exists in the correct location within the `Tycoon` folder.
- **Debugging**: Add print statements to confirm the existence of `MainItems` and its children.
Here you go

```lua
-- Services
local TweenService = game:GetService("TweenService")
local Players = game:GetService("Players")
local MarketplaceService = game:GetService("MarketplaceService")

-- References
local tycoon = script.Parent.Parent
local MainItems = tycoon:FindFirstChild("MainItems")
local values = tycoon:FindFirstChild("Values")
local buttons = tycoon:FindFirstChild("Buttons")
local PurchasedItems = tycoon:FindFirstChild("PurchasedItems")
local audio = tycoon:FindFirstChild("Audio")

-- Debugging references
if not MainItems then
    error("MainItems nie istnieje!")
end

if not MainItems:FindFirstChild("OwnerDoor") then
    error("OwnerDoor nie istnieje w MainItems!")
end

-- Variables
local objects = {}
local random = Random.new()
local BUILDING_ANIMATION_POSITION_OFFSET_AMOUNT = 2
local BUILDING_ANIMATION_PART_DELAY = 0.03

-- Helper functions
function hasProperty(instance, property)
    assert(typeof(instance) == "Instance")
    assert(typeof(property) == "string")

    local hasProperty = false
    pcall(function()
        local v = instance[property]
        hasProperty = true 
    end)

    return hasProperty
end

function instanceListToPropertyDict(instances, propertyList)
    assert(typeof(instances) == "table")
    assert(typeof(propertyList) == "table")

    local dict = {}
    for _, instance in ipairs(instances) do
        local dictEntry = {}
        for _, property in pairs(propertyList) do
            assert(hasProperty(instance, property), string.format("Instance '%s' (a %s) doesn't have property '%s'.", tostring(instance), instance.ClassName, property))
            dictEntry[property] = instance[property]
        end
        dict[instance] = dictEntry
    end
    return dict
end

function getDescendantsWhichAre(ancestor, className)
    assert(typeof(ancestor) == "Instance")
    assert(typeof(className) == "string")

    local descendants = {}
    for _, descendant in pairs(ancestor:GetDescendants()) do
        if descendant:IsA(className) then
            table.insert(descendants, descendant)
        end
    end
    return descendants
end

function animateBuildingIn(buildingModel, tweenInfo)
    assert(typeof(buildingModel) == "Instance" and buildingModel.ClassName == "Model", string.format("Invalid argument #1 to 'animateBuildingIn' (Model expected, got %s)", typeof(buildingModel)))
    assert(typeof(tweenInfo) == "TweenInfo", string.format("Invalid argument #2 to 'animateBuildingIn' (TweenInfo expected, got %s)", typeof(tweenInfo)))

    local parts = getDescendantsWhichAre(buildingModel, "BasePart")
    local originalProperties = instanceListToPropertyDict(parts, {"Transparency", "CFrame", "Color"})

    for _, part in pairs(parts) do
        part.Transparency = 1
        part.Color = Color3.fromRGB(255, 255, 255)

        local positionOffset = Vector3.new(random:NextNumber(-1, 1), random:NextNumber(-0.25, 1.75), random:NextNumber(-1, 1)) * BUILDING_ANIMATION_POSITION_OFFSET_AMOUNT
        part.CFrame = part.CFrame * CFrame.new(positionOffset)
    end

    for _, part in pairs(parts) do
        local tween = TweenService:Create(part, tweenInfo, originalProperties[part])

        tween.Completed:Connect(function()
            part.Transparency = originalProperties[part].Transparency
            part.CFrame = originalProperties[part].CFrame
        end)

        tween:Play()
        wait(BUILDING_ANIMATION_PART_DELAY)
    end
end

-- Main Logic - Tycoon Owner Door Functionality
MainItems.OwnerDoor.Door.Touched:Connect(function(hit)
    local player = Players:GetPlayerFromCharacter(hit.Parent)

    if player and values.OwnerValue.Value == nil then 
        if not player:FindFirstChild("HasTycoon") or player.HasTycoon.Value == false then 
            values.OwnerValue.Value = player 
            MainItems.OwnerDoor.Title.SurfaceGui.TextLabel.Text = tostring(values.OwnerValue.Value.Name) .. "'s Tycoon"
            player:SetAttribute("HasTycoon", true)
            print("Gracz stał się właścicielem Tycoona:", player.Name)  -- Debugging line 
        else 
            print("Gracz już posiada Tycoona.")
        end 
    else 
        print("Hit nie jest graczem lub Tycoon już ma właściciela.")
    end 
end)

if buttons then
    for _, v in pairs(buttons:GetChildren()) do
        if v:FindFirstChild("Button") then
            local objectName = v.Object.Value
            local priceValue = v.Price.Value

            spawn(function() 
                print("Przetwarzanie przycisku:", v.Name)  -- Debugging line 

                local newObject = PurchasedItems:FindFirstChild(objectName)
                if newObject then 
                    objects[objectName] = newObject:Clone()
                    newObject:Destroy() 
                    print("Skopiowany obiekt:", objectName)  -- Debugging line  
                else  
                    print("Nie znaleziono obiektu dla:", objectName)  -- Debugging line 
                    v:Destroy() 
                end 

                if v:FindFirstChild("Dependency") then 
                    v.Button.Transparency = 1 
                    v.Button.CanCollide = false 

                    coroutine.resume(coroutine.create(function() 
                        if PurchasedItems:FindFirstChild(v.Dependency.Value) then 
                            v.Button.Transparency = 0 
                            v.Button.CanCollide = true 
                        end 
                    end))  
                end  

                v.Button.Touched:Connect(function(hit) 
                    print("Przycisk dotknięty")  -- Debugging line 

                    local player = Players:GetPlayerFromCharacter(hit.Parent)

                    if player and values.OwnerValue.Value == player then 

                        if objects[objectName] then  
                            print("Gracz jest właścicielem")  -- Debugging line 

                            if v.Button.CanCollide then  
                                if player:FindFirstChild("LeaderStats") and player.LeaderStats.Cash.Value >= priceValue then  
                                    player.LeaderStats.Cash.Value = player.LeaderStats.Cash.Value - priceValue  

                                    -- Klonowanie obiektu Droppera i ustawianie go jako widocznego
                                    local clonedObject = objects[objectName]:Clone()  
                                    clonedObject.Parent = PurchasedItems  

                                    -- Ustawienie pozycji Droppera na położenie przycisku (zmień wartości według potrzeb)
                                    clonedObject:SetPrimaryPartCFrame(v.Button.Position + Vector3.new(0, v.Button.Size.Y / 2 + clonedObject.PrimaryPart.Size.Y / 2 + 1, 0)) -- Umieść Droppera nad przyciskiem

                                    -- Ustawienia widoczności Droppera
                                    clonedObject.PrimaryPart.Transparency = 0 -- Ustaw przezroczystość na 0 (widoczny)
                                    clonedObject.PrimaryPart.CanCollide = true -- Upewnij się, że koliduje

                                    print("Zakupiono:", objectName)  -- Debugging line 

                                    animateBuildingIn(clonedObject, TweenInfo.new(1, Enum.EasingStyle.Quad, Enum.EasingDirection.InOut))

                                    -- Pokaż przycisk kolorowania po zakupie Droppera i ukryj przycisk zakupu Droppera.
                                    local colorizerButton = MainItems:FindFirstChild("ColorizerButton")
                                    if colorizerButton then
                                        colorizerButton.Visible = true  
                                    end

                                    -- Ukryj przycisk zakupu po pierwszym zakupie.
                                    v.Button.Visible = false  
                                else  
                                    print("Niewystarczająca ilość pieniędzy lub brak LeaderStats")  -- Debugging line  
                                end  
                            else  
                                print("Przycisk nie może kolidować")  -- Debugging line  
                            end  
                        else  
                            print("Dropper już zakupiony.") -- Informacja o przekroczeniu limitu zakupu Droppera.
                        end  
                    else  
                        print("Gracz nie jest właścicielem lub hit jest nil.")  -- Debugging line  
                    end   
                end)   
            end)   
        end   
    end
end

PurchasedItems.ChildAdded:Connect(function(child)
    if child:IsA("Model") then
        print("Dziecko dodane do PurchasedItems:", child.Name) -- Debugging line

        task.delay(0.5, function() 
            animateBuildingIn(child, TweenInfo.new(1, Enum
